File: server/app/services/diff_service.py
"""
文件差异对比服务
支持文本文件和二进制文件的差异检测
"""
import difflib
import hashlib
from dataclasses import dataclass
from enum import Enum


class DiffType(Enum):
    ADD = "add"
    DELETE = "delete"
    MODIFY = "modify"
    EQUAL = "equal"


@dataclass
class DiffLine:
    line_number: int
    content: str
    diff_type: DiffType


@dataclass
class DiffResult:
    additions: int
    deletions: int
    changes: int
    from_content: str
    to_content: str
    from_content_html: str
    to_content_html: str
    changes_list: list[dict]


class FileDiffService:
    @staticmethod
    def get_file_hash(file_path: str) -> str:
        """计算文件哈希值"""
        hash_md5 = hashlib.md5()
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(4096), b""):
                hash_md5.update(chunk)
        return hash_md5.hexdigest()

    @staticmethod
    def is_text_file(file_path: str, sample_size: int = 8192) -> bool:
        """判断是否为文本文件"""
        try:
            with open(file_path, "rb") as f:
                chunk = f.read(sample_size)
                if b'\x00' in chunk:
                    return False
                try:
                    chunk.decode('utf-8')
                    return True
                except UnicodeDecodeError:
                    try:
                        chunk.decode('gbk')
                        return True
                    except UnicodeDecodeError:
                        return False
        except Exception:
            return False

    @staticmethod
    def read_file_content(file_path: str) -> tuple[list[str], str]:
        """读取文件内容,返回行列表和原始内容"""
        try:
            with open(file_path, encoding="utf-8") as f:
                content = f.read()
                lines = content.splitlines(keepends=True)
                return lines, content
        except UnicodeDecodeError:
            try:
                with open(file_path, encoding="gbk") as f:
                    content = f.read()
                    lines = content.splitlines(keepends=True)
                    return lines, content
            except Exception:
                return [], ""
        except Exception:
            return [], ""

    @staticmethod
    def compare_text_files(
        from_path: str,
        to_path: str,
        context_lines: int = 3
    ) -> DiffResult:
        """比较两个文本文件"""
        from_lines, from_content = FileDiffService.read_file_content(from_path)
        to_lines, to_content = FileDiffService.read_file_content(to_path)

        diff = difflib.unified_diff(
            from_lines,
            to_lines,
            fromfile="from",
            tofile="to",
            lineterm=""
        )

        additions = 0
        deletions = 0
        changes_list = []
        line_number = 0

        for line in diff:
            if line.startswith('+++') or line.startswith('---') or line.startswith('@@'):
                continue
            elif line.startswith('+'):
                additions += 1
                changes_list.append({
                    "type": "add",
                    "line": line_number,
                    "content": line[1:].strip()
                })
            elif line.startswith('-'):
                deletions += 1
                changes_list.append({
                    "type": "delete",
                    "line": line_number,
                    "content": line[1:].strip()
                })
            else:
                line_number += 1

        from_html = FileDiffService._generate_html_diff(from_lines, to_lines, is_from=True)
        to_html = FileDiffService._generate_html_diff(from_lines, to_lines, is_from=False)

        return DiffResult(
            additions=additions,
            deletions=deletions,
            changes=additions + deletions,
            from_content=from_content,
            to_content=to_content,
            from_content_html=from_html,
            to_content_html=to_html,
            changes_list=changes_list[:100]
        )

    @staticmethod
    def compare_binary_files(from_path: str, to_path: str) -> DiffResult:
        """比较两个二进制文件"""
        from_hash = FileDiffService.get_file_hash(from_path)
        to_hash = FileDiffService.get_file_hash(to_path)

        from_size = 0
        to_size = 0

        try:
            import os
            from_size = os.path.getsize(from_path)
            to_size = os.path.getsize(to_path)
        except Exception:
            pass

        if from_hash == to_hash:
            return DiffResult(
                additions=0,
                deletions=0,
                changes=0,
                from_content=f"Binary file (hash: {from_hash[:16]}..., size: {from_size} bytes)",
                to_content=f"Binary file (hash: {to_hash[:16]}..., size: {to_size} bytes)",
                from_content_html="<span style='color: gray'>Binary files are identical</span>",
                to_content_html="<span style='color: gray'>Binary files are identical</span>",
                changes_list=[]
            )

        size_diff = to_size - from_size
        additions = max(0, size_diff)
        deletions = max(0, -size_diff)

        return DiffResult(
            additions=additions,
            deletions=deletions,
            changes=1,
            from_content=f"Binary file (hash: {from_hash[:16]}..., size: {from_size} bytes)",
            to_content=f"Binary file (hash: {to_hash[:16]}..., size: {to_size} bytes)",
            from_content_html=f"<span style='color: red'>Binary file changed (hash: {from_hash[:16]}...)</span>",
            to_content_html=f"<span style='color: green'>Binary file changed (hash: {to_hash[:16]}...)</span>",
            changes_list=[{
                "type": "modify",
                "line": 0,
                "content": f"Binary file content changed, size difference: {size_diff} bytes"
            }]
        )

    @staticmethod
    def compare_files(from_path: str, to_path: str) -> DiffResult:
        """比较两个文件,自动判断文件类型"""
        if FileDiffService.is_text_file(from_path) and FileDiffService.is_text_file(to_path):
            return FileDiffService.compare_text_files(from_path, to_path)
        else:
            return FileDiffService.compare_binary_files(from_path, to_path)

    @staticmethod
    def _generate_html_diff(from_lines: list[str], to_lines: list[str], is_from: bool) -> str:
        """生成HTML格式的差异显示"""
        diff = list(difflib.ndiff(from_lines, to_lines))
        html_lines = []

        for line in diff:
            if line.startswith('  '):
                html_lines.append(f"<span>{line[2:]}</span>")
            elif line.startswith('- ') and is_from:
                html_lines.append(f"<span style='background-color: #ffecec; color: #a33'>{line[2:]}</span>")
            elif line.startswith('+ ') and not is_from:
                html_lines.append(f"<span style='background-color: #eaffea; color: #270'>{line[2:]}</span>")
            elif line.startswith('? '):
                pass

        return ''.join(html_lines[:500])

    @staticmethod
    def get_similarity(from_path: str, to_path: str) -> float:
        """计算文件相似度"""
        if not FileDiffService.is_text_file(from_path) or not FileDiffService.is_text_file(to_path):
            from_hash = FileDiffService.get_file_hash(from_path)
            to_hash = FileDiffService.get_file_hash(to_path)
            return 1.0 if from_hash == to_hash else 0.0

        from_lines, _ = FileDiffService.read_file_content(from_path)
        to_lines, _ = FileDiffService.read_file_content(to_path)

        if not from_lines and not to_lines:
            return 1.0
        if not from_lines or not to_lines:
            return 0.0

        matcher = difflib.SequenceMatcher(None, from_lines, to_lines)
        return matcher.ratio()


file_diff_service = FileDiffService()

File: apps/api/src/services/diff-service.ts
/**
 * 差异对比服务。
 * 支持文本文件（基于 LCS 的逐行 diff）与二进制文件（哈希对比）。
 * 零依赖实现，不引入外部 diff 库。
 */

import { createHash } from 'node:crypto';
import { readFileSync, existsSync, statSync } from 'node:fs';

export type DiffType = 'add' | 'delete' | 'modify' | 'equal';

export interface DiffLine {
  lineNumber: number;
  content: string;
  diffType: DiffType;
}

export interface DiffResult {
  additions: number;
  deletions: number;
  changes: number;
  fromContent: string;
  toContent: string;
  changesList: DiffLine[];
}

/** 计算文件 SHA-256 哈希。 */
export function getFileHash(filePath: string): string {
  const buf = readFileSync(filePath);
  return createHash('sha256').update(buf).digest('hex');
}

/** 判断是否为文本文件（采样前 8KB，含 \0 视为二进制）。 */
export function isTextFile(filePath: string, sampleSize = 8192): boolean {
  try {
    const fd = readFileSync(filePath, { encoding: null });
    const chunk = fd.subarray(0, sampleSize);
    if (chunk.includes(0)) return false;
    return Buffer.from(chunk).toString('utf8') !== '';
  } catch {
    return false;
  }
}

/** 读取文件内容，返回行数组与原始字符串。 */
export function readFileContent(filePath: string): { lines: string[]; content: string } {
  try {
    const content = readFileSync(filePath, 'utf-8');
    return { lines: content.split(/\r?\n/), content };
  } catch {
    return { lines: [], content: '' };
  }
}

/**
 * 基于 LCS 计算两段文本的逐行 diff。
 */
export function computeTextDiff(fromText: string, toText: string): DiffLine[] {
  const fromLines = fromText.split(/\r?\n/);
  const toLines = toText.split(/\r?\n/);
  const m = fromLines.length;
  const n = toLines.length;

  // LCS 动态规划表
  const dp: number[][] = Array.from({ length: m + 1 }, () => new Array<number>(n + 1).fill(0));
  for (let i = m - 1; i >= 0; i--) {
    const row = dp[i]!;
    const nextRow = dp[i + 1]!;
    for (let j = n - 1; j >= 0; j--) {
      row[j] = fromLines[i] === toLines[j] ? (nextRow[j + 1] ?? 0) + 1 : Math.max(nextRow[j] ?? 0, row[j + 1] ?? 0);
    }
  }

  const result: DiffLine[] = [];
  let i = 0;
  let j = 0;
  let lineNum = 0;
  while (i < m && j < n) {
    const row = dp[i]!;
    const nextRow = dp[i + 1]!;
    if (fromLines[i] === toLines[j]) {
      result.push({ lineNumber: lineNum++, content: fromLines[i]!, diffType: 'equal' });
      i++;
      j++;
    } else if ((nextRow[j] ?? 0) >= (row[j + 1] ?? 0)) {
      result.push({ lineNumber: lineNum, content: fromLines[i]!, diffType: 'delete' });
      i++;
    } else {
      result.push({ lineNumber: lineNum++, content: toLines[j]!, diffType: 'add' });
      j++;
    }
  }
  while (i < m) result.push({ lineNumber: lineNum, content: fromLines[i++]!, diffType: 'delete' });
  while (j < n) result.push({ lineNumber: lineNum++, content: toLines[j++]!, diffType: 'add' });
  return result;
}

/** 比较两个文本文件。 */
export function compareTextFiles(fromPath: string, toPath: string): DiffResult {
  const { content: fromContent } = readFileContent(fromPath);
  const { content: toContent } = readFileContent(toPath);

  const diffLines = computeTextDiff(fromContent, toContent);
  const additions = diffLines.filter((d) => d.diffType === 'add').length;
  const deletions = diffLines.filter((d) => d.diffType === 'delete').length;

  return {
    additions,
    deletions,
    changes: additions + deletions,
    fromContent,
    toContent,
    changesList: diffLines.filter((d) => d.diffType !== 'equal').slice(0, 100),
  };
}

/** 比较两个二进制文件（基于哈希）。 */
export function compareBinaryFiles(fromPath: string, toPath: string): DiffResult {
  const fromHash = getFileHash(fromPath);
  const toHash = getFileHash(toPath);
  const fromSize = existsSync(fromPath) ? statSync(fromPath).size : 0;
  const toSize = existsSync(toPath) ? statSync(toPath).size : 0;

  if (fromHash === toHash) {
    return {
      additions: 0,
      deletions: 0,
      changes: 0,
      fromContent: `Binary (hash: ${fromHash.slice(0, 16)}, size: ${fromSize})`,
      toContent: `Binary (hash: ${toHash.slice(0, 16)}, size: ${toSize})`,
      changesList: [],
    };
  }

  const sizeDiff = toSize - fromSize;
  return {
    additions: Math.max(0, sizeDiff),
    deletions: Math.max(0, -sizeDiff),
    changes: 1,
    fromContent: `Binary (hash: ${fromHash.slice(0, 16)}, size: ${fromSize})`,
    toContent: `Binary (hash: ${toHash.slice(0, 16)}, size: ${toSize})`,
    changesList: [
      { lineNumber: 0, content: `Binary changed, size diff: ${sizeDiff} bytes`, diffType: 'modify' },
    ],
  };
}

/** 自动判断文件类型并比较。 */
export function compareFiles(fromPath: string, toPath: string): DiffResult {
  if (isTextFile(fromPath) && isTextFile(toPath)) {
    return compareTextFiles(fromPath, toPath);
  }
  return compareBinaryFiles(fromPath, toPath);
}

/** 计算两文件相似度（0-1）。 */
export function getSimilarity(fromPath: string, toPath: string): number {
  if (!isTextFile(fromPath) || !isTextFile(toPath)) {
    return getFileHash(fromPath) === getFileHash(toPath) ? 1.0 : 0.0;
  }
  const { lines: fromLines } = readFileContent(fromPath);
  const { lines: toLines } = readFileContent(toPath);
  if (fromLines.length === 0 && toLines.length === 0) return 1.0;
  if (fromLines.length === 0 || toLines.length === 0) return 0.0;

  // LCS 长度 / 较大行数
  const m = fromLines.length;
  const n = toLines.length;
  const dp: number[][] = Array.from({ length: m + 1 }, () => new Array<number>(n + 1).fill(0));
  for (let i = m - 1; i >= 0; i--) {
    const row = dp[i]!;
    const nextRow = dp[i + 1]!;
    for (let j = n - 1; j >= 0; j--) {
      row[j] = fromLines[i] === toLines[j] ? (nextRow[j + 1] ?? 0) + 1 : Math.max(nextRow[j] ?? 0, row[j + 1] ?? 0);
    }
  }
  return (dp[0]?.[0] ?? 0) / Math.max(m, n);
}
